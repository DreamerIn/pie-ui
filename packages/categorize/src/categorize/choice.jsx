import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import { DragSource } from 'react-dnd';
import { uid } from '@pie-lib/drag';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import green from '@material-ui/core/colors/green';
import orange from '@material-ui/core/colors/orange';
import debug from 'debug';

const log = debug('@pie-ui:categorize:choice');

export const ChoiceType = {
  content: PropTypes.string.isRequired,
  id: PropTypes.string
};

export class Layout extends React.Component {
  static propTypes = {
    ...ChoiceType,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    correct: PropTypes.bool
  };
  static defaultProps = {};
  render() {
    const {
      classes,
      className,
      content,
      isDragging,
      disabled,
      correct
    } = this.props;

    const rootNames = classNames(
      correct === true && 'correct',
      correct === false && 'incorrect',
      classes.choice,
      isDragging && classes.dragging,
      disabled && classes.disabled,
      className
    );
    const cardNames = classNames(classes.card);
    return (
      <div className={rootNames}>
        <Card className={cardNames}>
          <CardContent dangerouslySetInnerHTML={{ __html: content }} />
        </Card>
      </div>
    );
  }
}

const styles = () => ({
  choice: {
    cursor: 'pointer',
    height: '100%',
    width: '100%',
    '&.correct': {
      border: `solid 1px ${green[500]}`
    },
    '&.incorrect': {
      border: `solid 1px ${orange[500]}`
    }
  },
  disabled: {
    cursor: 'inherit'
  },
  dragging: {
    cursor: 'move'
  },
  card: {
    width: '100%'
  }
});

const Styled = withStyles(styles)(Layout);

export class Choice extends React.Component {
  static propTypes = {
    ...ChoiceType,
    connectDragSource: PropTypes.func.isRequired
  };

  render() {
    const {
      connectDragSource,
      id,
      content,
      disabled,
      isDragging,
      correct
    } = this.props;
    return connectDragSource(
      <div style={{ width: '100%', height: '100%' }}>
        <Styled
          id={id}
          content={content}
          disabled={disabled}
          correct={correct}
          isDragging={isDragging}
        />
      </div>
    );
  }
}

export const spec = {
  canDrag: props => !props.disabled,
  beginDrag: props => {
    const out = {
      id: props.id,
      categoryId: props.categoryId,
      choiceIndex: props.choiceIndex
    };
    log('[beginDrag] out:', out);
    return out;
  },
  endDrag: (props, monitor) => {
    if (!monitor.didDrop()) {
      const item = monitor.getItem();
      if (item.categoryId) {
        log('wasnt droppped - what to do?');
        props.onRemoveChoice(item);
      }
    }
  }
};

const DraggableChoice = DragSource(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  })
)(Choice);

export default uid.withUid(DraggableChoice);
