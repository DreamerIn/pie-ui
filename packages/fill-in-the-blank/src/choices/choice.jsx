import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import { DragSource } from 'react-dnd';
import debug from 'debug';
import { uid } from '@pie-lib/drag';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import grey from '@material-ui/core/colors/grey';
import orange from '@material-ui/core/colors/orange';
import green from '@material-ui/core/colors/green';

const log = debug('@pie-ui:fill-in-the-blank:choice');

export class Choice extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    content: PropTypes.string,
    connectDragSource: PropTypes.func.isRequired,
    blankId: PropTypes.string,
    disabled: PropTypes.bool,
    correct: PropTypes.bool
  };
  static defaultProps = {};
  render() {
    const {
      classes,
      className,
      content,
      connectDragSource,
      disabled,
      correct
    } = this.props;

    return connectDragSource(
      <div
        className={classNames(
          classes.choice,
          disabled && classes.disabledChoice,
          className
        )}
      >
        <Card
          className={classNames(
            correct === false && classes.incorrect,
            correct === true && classes.correct,
            disabled && correct === undefined && classes.disabledCard
          )}
        >
          <CardContent>
            <div dangerouslySetInnerHTML={{ __html: content }} />
          </CardContent>
        </Card>
      </div>
    );
  }
}
const styles = theme => ({
  choice: {
    padding: 0,
    backgroundColor: 'white',
    cursor: 'move'
  },
  disabledChoice: {
    cursor: 'auto'
  },
  incorrect: {
    border: `solid 1px ${orange[500]}`,
    boxShadow: 'none',
    borderRadius: 0
  },
  correct: {
    border: `solid 1px ${green[500]}`,
    boxShadow: 'none',
    borderRadius: 0
  },
  disabledCard: {
    boxShadow: 'none',
    border: `solid 1px ${grey[400]}`
  }
});

const Styled = withStyles(styles)(Choice);

export const spec = {
  canDrag: props => !props.disabled,
  beginDrag: props => {
    const out = {
      id: props.id,
      blankId: props.blankId
    };
    log('[beginDrag] out:', out);
    return out;
  },
  endDrag: (props, monitor) => {
    if (!monitor.didDrop()) {
      log('did not drop?');
      const item = monitor.getItem();

      if (item.blankId) {
        log('wasnt droppped - what to do?');
        if (props.onRemoveFromBlank) {
          props.onRemoveFromBlank(item);
        }
      }
    }
  }
};

const Draggable = DragSource(
  ({ uid }) => {
    log('uid: ', uid);
    return uid;
  },
  spec,
  (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  })
)(Styled);

export default uid.withUid(Draggable);
