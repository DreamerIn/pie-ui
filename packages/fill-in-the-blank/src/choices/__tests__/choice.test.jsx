import { shallow } from 'enzyme';
import React from 'react';
import { Choice, spec } from '../choice';

describe('Choice', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      content: 'hi',
      id: '1',
      className: 'className',
      onChange,
      connectDragSource: jest.fn(n => n)
    };

    const props = { ...defaults, ...extras };
    return shallow(<Choice {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('spec', () => {
    describe('canDrag', () => {
      it('returns true if !disabled', () => {
        expect(spec.canDrag({ disabled: false })).toEqual(true);
      });
      it('returns true if !disabled', () => {
        expect(spec.canDrag({ disabled: true })).toEqual(false);
      });
    });

    describe('beginDrag', () => {
      it('returns id and blankId', () => {
        expect(spec.beginDrag({ id: '1', blankId: '1' })).toEqual({
          id: '1',
          blankId: '1'
        });
      });
    });

    describe('endDrag', () => {
      let props, monitor, item;

      beforeEach(() => {
        item = {
          id: '1',
          blankId: '1'
        };
        monitor = {
          didDrop: jest.fn(() => false),
          getItem: jest.fn(() => item)
        };

        props = {
          onRemoveFromBlank: jest.fn()
        };
      });
      it('calls onRemoveFromBlank', () => {
        spec.endDrag(props, monitor);
        expect(props.onRemoveFromBlank).toBeCalledWith(item);
      });
      it('does not call onRemoveFromBlank if there is no blankId', () => {
        item.blankId = undefined;
        spec.endDrag(props, monitor);
        expect(props.onRemoveFromBlank).not.toBeCalled();
      });
      it('does not call onRemoveFromBlank if didDrop returns true', () => {
        monitor.didDrop = jest.fn(() => true);
        spec.endDrag(props, monitor);
        expect(props.onRemoveFromBlank).not.toBeCalled();
      });
    });
  });
});
