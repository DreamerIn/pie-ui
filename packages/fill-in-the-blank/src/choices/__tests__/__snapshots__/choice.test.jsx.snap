// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice snapshot renders 1`] = `
<div
  className="className"
>
  <WithStyles(Card)
    className=""
  >
    <WithStyles(CardContent)>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "hi",
          }
        }
      />
    </WithStyles(CardContent)>
  </WithStyles(Card)>
</div>
`;
